// @aws-cdk/toolkit-lib/lib/actions/deploy/private/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildParameterMap = buildParameterMap;
exports.removePublishedAssetsFromWorkGraph = removePublishedAssetsFromWorkGraph;
exports.createHotswapPropertyOverrides = createHotswapPropertyOverrides;
const shared_private_1 = require("../../../api/shared-private");
function buildParameterMap(parameters) {
    const parameterMap = {};
    parameterMap['*'] = {};
    const entries = parameters?.entries() ?? [];
    for (const [key, value] of entries) {
        const [stack, parameter] = key.split(':', 2);
        if (!parameter) {
            parameterMap['*'][stack] = value;
        }
        else {
            if (!parameterMap[stack]) {
                parameterMap[stack] = {};
            }
            parameterMap[stack][parameter] = value;
        }
    }
    return parameterMap;
}
/**
 * Remove the asset publishing and building from the work graph for assets that are already in place
 */
async function removePublishedAssetsFromWorkGraph(graph, deployments, options) {
    await graph.removeUnnecessaryAssets(assetNode => deployments.isSingleAssetPublished(assetNode.assetManifest, assetNode.asset, {
        stack: assetNode.parentStack,
        roleArn: options.roleArn,
        stackName: assetNode.parentStack.stackName,
    }));
}
/**
 * Create the HotswapPropertyOverrides class out of the Interface exposed to users
 */
function createHotswapPropertyOverrides(hotswapProperties) {
    return new shared_private_1.HotswapPropertyOverrides(new shared_private_1.EcsHotswapProperties(hotswapProperties.ecs.minimumHealthyPercent, hotswapProperties.ecs.maximumHealthyPercent));
}
