// @aws-cdk/cloudformation-diff/lib/mappings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatTypedMappings = formatTypedMappings;
exports.formatAmbiguousMappings = formatAmbiguousMappings;
const chalk = require("chalk");
const format_1 = require("./format");
const format_table_1 = require("./format-table");
function formatTypedMappings(stream, mappings) {
    const header = [['Resource Type', 'Old Construct Path', 'New Construct Path']];
    const rows = mappings.map((m) => [m.type, m.sourcePath, m.destinationPath]);
    const formatter = new format_1.Formatter(stream, {});
    if (mappings.length > 0) {
        formatter.printSectionHeader('The following resources were moved or renamed:');
        formatter.print(chalk.green((0, format_table_1.formatTable)(header.concat(rows), undefined)));
    }
    else {
        formatter.print('Nothing to refactor.');
    }
}
function formatAmbiguousMappings(stream, pairs) {
    const tables = pairs.map(renderTable);
    const formatter = new format_1.Formatter(stream, {});
    formatter.printSectionHeader('Ambiguous Resource Name Changes');
    formatter.print(tables.join('\n\n'));
    formatter.printSectionFooter();
    function renderTable([removed, added]) {
        return (0, format_table_1.formatTable)([['', 'Resource'], renderRemoval(removed), renderAddition(added)], undefined);
    }
    function renderRemoval(locations) {
        return [chalk.red('-'), chalk.red(renderLocations(locations))];
    }
    function renderAddition(locations) {
        return [chalk.green('+'), chalk.green(renderLocations(locations))];
    }
    function renderLocations(locs) {
        return locs.join('\n');
    }
}
