// @aws-cdk/toolkit-lib/lib/api/cloud-assembly/private/exec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execInChildProcess = execInChildProcess;
const child_process = require("node:child_process");
// eslint-disable-next-line @typescript-eslint/no-require-imports
const split = require("split2");
const shared_public_1 = require("../../shared-public");
/**
 * Execute a command and args in a child process
 */
async function execInChildProcess(commandAndArgs, options = {}) {
    return new Promise((ok, fail) => {
        // We use a slightly lower-level interface to:
        //
        // - Pass arguments in an array instead of a string, to get around a
        //   number of quoting issues introduced by the intermediate shell layer
        //   (which would be different between Linux and Windows).
        //
        // - We have to capture any output to stdout and stderr sp we can pass it on to the IoHost
        //   To ensure messages get to the user fast, we will emit every full line we receive.
        const proc = child_process.spawn(commandAndArgs, {
            stdio: ['ignore', 'pipe', 'pipe'],
            detached: false,
            shell: true,
            cwd: options.cwd,
            env: {
                ...process.env,
                ...(options.extraEnv ?? {}),
            },
        });
        const eventPublisher = options.eventPublisher ?? ((type, line) => {
            switch (type) {
                case 'data_stdout':
                    process.stdout.write(line);
                    return;
                case 'data_stderr':
                    process.stderr.write(line);
                    return;
                case 'open':
                case 'close':
                    return;
            }
        });
        proc.stdout.pipe(split()).on('data', (line) => eventPublisher('data_stdout', line));
        proc.stderr.pipe(split()).on('data', (line) => eventPublisher('data_stderr', line));
        proc.on('error', fail);
        proc.on('exit', code => {
            if (code === 0) {
                return ok();
            }
            else {
                return fail(new shared_public_1.ToolkitError(`Subprocess exited with error ${code}`));
            }
        });
    });
}
