// @aws-cdk/toolkit-lib/lib/actions/diff/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiffMethod = void 0;
class DiffMethod {
    method;
    options;
    /**
     * Use a changeset to compute the diff.
     *
     * This will create, analyze, and subsequently delete a changeset against the CloudFormation stack.
     */
    static ChangeSet(options = {}) {
        return new class extends DiffMethod {
            options;
            constructor(opts) {
                super('change-set', opts);
                this.options = opts;
            }
        }(options);
    }
    static TemplateOnly(options = {}) {
        return new class extends DiffMethod {
            options;
            constructor(opts) {
                super('template-only', opts);
                this.options = opts;
            }
        }(options);
    }
    /**
     * Use a local template file to compute the diff.
     */
    static LocalFile(path) {
        return new class extends DiffMethod {
            options;
            constructor(opts) {
                super('local-file', opts);
                this.options = opts;
            }
        }({ path });
    }
    constructor(method, options) {
        this.method = method;
        this.options = options;
    }
}
exports.DiffMethod = DiffMethod;
