// cdk-assets/lib/progress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventType = void 0;
exports.shellEventToEventType = shellEventToEventType;
exports.shellEventPublisherFromEventEmitter = shellEventPublisherFromEventEmitter;
/**
 * A single event for an asset
 */
var EventType;
(function (EventType) {
    /**
     * Just starting on an asset
     */
    EventType["START"] = "start";
    /**
     * When an asset is successfully finished
     */
    EventType["SUCCESS"] = "success";
    /**
     * When an asset failed
     */
    EventType["FAIL"] = "fail";
    /**
     * Checking whether an asset has already been published
     */
    EventType["CHECK"] = "check";
    /**
     * The asset was already published
     */
    EventType["FOUND"] = "found";
    /**
     * The asset was reused locally from a cached version
     */
    EventType["CACHED"] = "cached";
    /**
     * The asset will be built
     */
    EventType["BUILD"] = "build";
    /**
     * The asset will be uploaded
     */
    EventType["UPLOAD"] = "upload";
    /**
     * Another type of detail message
     */
    EventType["DEBUG"] = "debug";
    /**
     * When a shell command is executed. Emits the the command line arguments given to
     * the subprocess as a string upon shell execution.
     *
     * Only emitted when subprocessOutputDestination is set to 'publish'
     */
    EventType["SHELL_OPEN"] = "shell_open";
    /**
     * stdout from a shell command
     *
     * Only emitted when subprocessOutputDestination is set to 'publish'
     */
    EventType["SHELL_STDOUT"] = "shell_stdout";
    /**
     * stdout from a shell command
     *
     * Only emitted when subprocessOutputDestination is set to 'publish'
     */
    EventType["SHELL_STDERR"] = "shell_stderr";
    /**
     * When a shell command closes. Emits the the command line arguments given to
     * the subprocess as a string upon shell closure.
     *
     * Only emitted when subprocessOutputDestination is set to 'publish'
     */
    EventType["SHELL_CLOSE"] = "shell_close";
})(EventType || (exports.EventType = EventType = {}));
/**
 * A helper function to convert shell events to asset progress events
 * @param event a shell event
 * @returns an {@link EventType}
 */
function shellEventToEventType(event) {
    switch (event) {
        case 'open':
            return EventType.SHELL_OPEN;
        case 'close':
            return EventType.SHELL_CLOSE;
        case 'data_stdout':
            return EventType.SHELL_STDOUT;
        case 'data_stderr':
            return EventType.SHELL_STDERR;
    }
}
/**
 * Create a ShellEventPublisher for an {@link EventEmitter}
 *
 * Returns a ShellEventPublisher that translates ShellEvents into messages
 * emitted via the EventEmitter. Uses `shellEventToEventType` to translate
 * shell event types to event types
 */
function shellEventPublisherFromEventEmitter(eventEmitter) {
    return (event, message) => {
        const eventType = shellEventToEventType(event);
        eventEmitter(eventType, message);
    };
}
